mock/mock_proxy.go

```diff
@@ -0,0 +1,280 @@
+
+	lastStdinStream []byte
+
+	Signal           chan ShimSignal
+	ShimDisconnected chan bool
+	StdinReceived    chan bool
+}
+
+func NewMockProxy(t *testing.T, path string) *MockProxy {
+	return &MockProxy{
+		t:                t,
+		connectionPath:   path,
+		lastStdinStream:  nil,
+		Signal:           make(chan ShimSignal, 5),
+		ShimConnected:    make(chan bool),
+		ShimDisconnected: make(chan bool),
+		StdinReceived:    make(chan bool),
+	}
+}
+
+func newSignalList() []ShimSignal {
+	return make([]ShimSignal, 0, 5)
+}
+
+func (proxy *MockProxy) GetLastStdinStream() []byte {
+	return proxy.lastStdinStream
+}
+
+func (proxy *MockProxy) log(s string) {
+	proxy.logF("%s\n", s)
+}
+
+func (proxy *MockProxy) logF(format string, args ...interface{}) {
+	proxy.t.Logf("[Proxy] "+format, args...)
+}
+
+type client struct {
+	proxy *MockProxy
+	conn  net.Conn
+}
+
+// Defining the payload here, as it has not been defined in proxy api package
+// yet
+type ConnectShim struct {
+	Token string `json:"token"`
+}
+
+type ShimSignal struct {
+	Signal int `json:"signal"`
+	Row    int `json:"row,omitempty"`
+	Column int `json:"column,omitempty"`
+}
+
+func connectShimHandler(data []byte, userData interface{}, response *handlerResponse) {
+	client := userData.(*client)
+	proxy := client.proxy
+
+	payload := ConnectShim{}
+	err := json.Unmarshal(data, &payload)
+	assert.Nil(proxy.t, err)
+
+	assert.Equal(proxy.t, payload.Token, testToken)
+
+	proxy.logF("ConnectShim(token=%s)", payload.Token)
+
+	response.AddResult("version", api.Version)
+	proxy.ShimConnected <- true
+}
+
+func signalShimHandler(data []byte, userData interface{}, response *handlerResponse) {
+	client := userData.(*client)
+	proxy := client.proxy
+
+	signalPayload := ShimSignal{}
+	err := json.Unmarshal(data, &signalPayload)
+	assert.Nil(proxy.t, err)
+
+	proxy.logF("Proxy received signal: %v", signalPayload)
+
+	proxy.Signal <- signalPayload
+}
+
+func disconnectShimHandler(data []byte, userData interface{}, response *handlerResponse) {
+	client := userData.(*client)
+	proxy := client.proxy
+
+	proxy.log("Client sent DisconnectShim Command")
+	proxy.ShimDisconnected <- true
+}
+
+func stdinShimHandler(data []byte, userData interface{}, response *handlerResponse) {
+	client := userData.(*client)
+	proxy := client.proxy
+
+	proxy.lastStdinStream = data
+	proxy.StdinReceived <- true
+}
+
+func (proxy *MockProxy) SendStdoutStream(payload []byte) {
+	err := api.WriteStream(proxy.cl, api.StreamStdout, payload)
+	assert.Nil(proxy.t, err)
+}
+
+func (proxy *MockProxy) SendStderrStream(payload []byte) {
+	err := api.WriteStream(proxy.cl, api.StreamStderr, payload)
+	assert.Nil(proxy.t, err)
+}
+
+func (proxy *MockProxy) SendExitNotification(payload []byte) {
+	err := api.WriteNotification(proxy.cl, api.NotificationProcessExited, payload)
+	assert.Nil(proxy.t, err)
+}
+
+func (proxy *MockProxy) startListening() {
+
+	l, err := net.ListenUnix("unix", &net.UnixAddr{Name: proxy.connectionPath, Net: "unix"})
```

Comment:
And maybe use an URL instead of a connection path. Your connection path will be embedded inside the URL.